packages/adapter-netlify/README.md

```diff
@@ -27,13 +27,12 @@ export default {
 };
 ```
 
-Then, make sure you have a [netlify.toml](https://docs.netlify.com/configure-builds/file-based-configuration) file in the project root. This will determine where to write static assets and functions to based on the `build.publish` and `build.functions` settings, as per this sample configuration:
+Then, make sure you have a [netlify.toml](https://docs.netlify.com/configure-builds/file-based-configuration) file in the project root. This will determine where to write static assets and functions to based on the `build.publish` settings, as per this sample configuration:
 
 ```toml
 [build]
   command = "npm run build"
   publish = "build/publish/"
-  functions = "build/functions/"
```

Comment:
I don't know if we can just remove this. Where will it put functions then? We put this here on purpose so that all output would show up under `build` and be automatically gitignored without the user having to edit their `.gitignore` at all. It would be nice if that continued to work for people who are not using the Netlify CLI

Reply:
Or at least if we're going to remove this line, should we also remove the `publish` line so that ends up under `.netlify/publish` or something? It seems weird to treat these inconsistently.

Perhaps what we could do is maintain the old behavior and use the directories in these files if they are present and if they are not then we use the directory you're specifying as a default

Reply:
The functions will be in `.netlify/functions-internal`. The `functions` directory isn't needed anymore, because we're using the internal functions directory instead, which is hard-coded. The idea is that the `functions` value is for user functions, while `.netlify/functions-internal` is for auto-generated functions. If we remove the `publish` dir it will default to `static` because that's [the value specified for SvelteKit](https://github.com/netlify/framework-info/blob/main/src/frameworks/svelte-kit.json#L17). Should that be changed?

Reply:
If they user isn't using the Netlify CLI then presumably this won't be an issue anyway, as it wouldn't be using the Netlify preset when building locally as the auto-detection won't trigger it.

Reply:
Though I suppose the user could have manually specified the preset.

Reply:
`static` seems a very odd place to put it. I would recommend we change https://github.com/netlify/framework-info/blob/16d5b7ea82c5c1d783e3711d2f8f1adf2da7eecb/src/frameworks/svelte-kit.json#L17 to something else. Either something under `build` or `.netlify` would probably make the most sense

I don't know what "preset" refers to

Up until now, we'd advised to put the following in `netlify.toml`:
```
  publish = "build/publish/"
  functions = "build/functions/"
```

The idea was that all adapters should put their output under `build` so that it was located in a standard place and already included in the `.gitignore` which includes `build` by default. I'm not going to use the Netlify CLI since I don't want any Netlify-specific functionality in my project, so I don't necessarily want to update my project and have it start putting the output someplace new and lose the ability to control where the output is located

Reply:
I'm not sure why framework-info has `static` as the default. Was it previously the output location? It is supposed to reflect each framework's default build location. We should certainly change it to whatever is the default.

I understand the desire to have everything under `build`, but the problem with having a user-specified functions directory is as #1249 shows, it means that there's no way for users to create their own functions. This need to separate user-generated and framework-generated functions was what led us to create the internal functions directory. 

Is there a scenario where somebody would be manually specifying the `netlify` adapter, but running `svelte-kit build` locally without the Netlify CLI? I'd assume that anybody using the Netlify adapter would either be running the build in CI, or if building locally they would be using the Netlify CLI to deploy.

Reply:
> Was it previously the output location?

`static` was never the output location. It's typically where you'd put your images, etc. When you do `npm init svelte@next` it will create a new project with a `static` folder that has a few image files in it used by the templates that live in `src/routes`. We should change it to be `build` or something.

> I understand the desire to have everything under build, but the problem with having a user-specified functions directory is as #1249 shows, it means that there's no way for users to create their own functions

I tried reading the Netlify docs on this and just got more confused. The `adapter-netlify` docs currently show:
```
[build]
  functions = "build/functions/"
```

How does Netlify know that the functions are in `build/functions`? Does it read that config value? If so, it appears to be undocumented

[What Netlify shows](https://docs.netlify.com/configure-builds/file-based-configuration/#functions) is:
```
[functions]
  directory =
```

> Is there a scenario where somebody would be manually specifying the netlify adapter, but running svelte-kit build locally without the Netlify CLI?

You have to do `npm run build` in order to do `npm run preview`, which is useful locally

Reply:
> > Was it previously the output location?
> 
> `static` was never the output location. It's typically where you'd put your images, etc. When you do `npm init svelte@next` it will create a new project with a `static` folder that has a few image files in it used by the templates that live in `src/routes`. We should change it to be `build` or something: [netlify/framework-info#380](https://github.com/netlify/framework-info/pull/380)

Wouldn't `build/publish` be the correct value?

> I tried reading the Netlify docs on this and just got more confused. The `adapter-netlify` docs currently show:
> 
> ```
> [build]
>   functions = "build/functions/"
> ```
> 
> How does Netlify know that the functions are in `build/functions`? Does it read that config value? If so, it appears to be undocumented

`build.functions` is the deprecated config syntax. The preferred syntax now is `functions.directory`, though the old syntax still works.
> 
> > Is there a scenario where somebody would be manually specifying the netlify adapter, but running svelte-kit build locally without the Netlify CLI?
> 
> You have to do `npm run build` in order to do `npm run preview`, which is useful locally

Reply:
How about this: we keep the default output in build, but remove the doc specifying the functions config. Then we detect if the build command is being run via the netlify cli, and if it is then either change the output to `.netlify/functions-internal` or copy the built files over. That way local builds without the CLI don't generate non-gitignored files, but builds in CI or via the cli get put in the correct place. Meanwhile users are free to create their own Netlify functions and specify whichever functions directory they want.

Reply:
OK, I've been thinking about this more, and my suggestion is:

- Set `publish` in `netlify.toml` to `build`, and put all the static and client assets there
- Do not set `functions` in `netlify.toml`. If the user has their own functions let them set it
- Set the esbuild outfile to `.svelte-kit/netlify/__render.js`
- When building, if we either detect that we're running via the Netlify CLI (we can do this via env vars), or if `.netlify/state.json` exists (which indicates they've already used the Netlify CLI locally), then copy `.svelte-kit/netlify/__render.js` to `.netlify/functions-internal/__render.js`

If you're happy with this then I can update the PR.

Reply:
What environment variable does the Netlify CLI set?

Reply:
In CI it sets `NETLIFY` and locally it sets `NETLIFY_LOCAL`

Reply:
ok, that plan basically sounds good to me. we could probably just look for the environment variable and don't need to look for `.netlify/state.json` since it might not always exist

maybe if the environment variable is not present then we can keep it working the same as it does today where we read the locations from the `netlify.toml` so folks not using the netlify CLI can put their build wherever they want

Reply:
> ok, that plan basically sounds good to me. we could probably just look for the environment variable and don't need to look for `.netlify/state.json` since it might not always exist

The scenario I'm thinking about when checking for `.netlify/state.json` as well as the env vars is where somebody who is using the Netlify CLI to deploy (i.e. just to do the upload), but is calling the build command directly rather than via `netlify build`. In that scenario we wouldn't move the functions into the right place, so when they deploy they'd be missing. If we check for the `state.json` we'd catch those cases and build them to the correct place for upload, and anyone who has that will already have the directory in their `.gitignore`.

> maybe if the environment variable is not present then we can keep it working the same as it does today where we read the locations from the `netlify.toml` so folks not using the netlify CLI can put their build wherever they want

This would mean it wouldn't fix #1249 for those people. It could also catch people who've created custom Netlify functions, specified a directory, and then run a build without the Netlify CLI. In that scenario it would delete all of their custom functions. 😬 

I think it's more consistent to just put them in the `.svelte-kit` subdirectory, where there's no chance of it overwriting user code.

Reply:
> The scenario I'm thinking about when checking for .netlify/state.json as well as the env vars is where somebody who is using the Netlify CLI to deploy (i.e. just to do the upload), but is calling the build command directly rather than via netlify build

I don't understand this scenario. If you built with `npm run build` then no `.netlify/state.json` will be created. How can we know that file will exist? Does that file get checked in? (I was assuming the whole `.netlify` directory is `.gitignore`d)

> It could also catch people who've created custom Netlify functions, specified a directory, and then run a build without the Netlify CLI. 

I don't understand this scenario either. They'd have had to have specify the output directory as the location where their source code resides. Why would anyone do that?

>I think it's more consistent to just put them in the .svelte-kit subdirectory, where there's no chance of it overwriting user code.

What is "them" in this sentence? The generated functions?

Reply:
> I don't understand this scenario. If you built with `npm run build` then no `.netlify/state.json` will be created. How can we know that file will exist? Does that file get checked in? (I was assuming the whole `.netlify` directory is `.gitignore`d)
> 

If you want to use the Netlify CLI in a local directory you have to first run `netlify link` to associate it with a site. This generates `.netlify/state.json` and adds `.netlify` to `.gitignore`.


> I don't understand this scenario either. They'd have had to have specify the output directory as the location where their source code resides. Why would anyone do that?

I thought that's what you were referring to: i.e. the functions directory? For most Netlify sites that directory contains source code for Netlify functions, but the current SvelteKit behavior (that we're changing here) is to write the generated functions there. Or are you referring to the `publish` directory? If so, then yes I think we'd read that from the toml in all cases, and default to `build` if it's not specified.

> 
> > I think it's more consistent to just put them in the .svelte-kit subdirectory, where there's no chance of it overwriting user code.
> 
> What is "them" in this sentence? The generated functions?

Yes

Reply:
Ok, thanks for clarifying and being patient as I come up-to-speed on Netlify stuff. I'm on board with your plan. Lets always output to `.netlify/functions-internal`

Reply:
OK, so skip the CLI detection entirely?

Reply:
What would be the benefit of CLI detection? If we're not using the Netlify CLI, I'm assuming that any code `.netlify/functions-internal` will still be used?

Reply:
Correct. I'll update the PR.

Reply:
OK, I've updated the PR as follows:

- Always write out the generated functions to `.netlify/functions-internal`
- Read `"publish"` from `netlify.toml`, but if it's missing then default to `"build"`
- Add more warnings and checks, such as ensuring that publish isn't set to the site root (that would delete the site!)
- Updated the README

One thing I'd like to suggest for a follow-up PR is that `cwd` is passed to adapter functions. It's already available in `adapt`, and it would ensure that we resolve all the paths correctly inside adapters. As it stands I'm not sure if these will all work if `cwd` isn't `process.cwd()`